### PR TITLE
Refactor synchronized_received_certificates_from_validator

### PR DESCRIPTION
**Pull request by deuszx**
## Motivation

Working on adding `download_certificates` endpoint and as part of the work I refactored `synchronized_received_certificates_from_validator` method a bit. Posting this small change as a separate PR to simplify and speedup discussions.

## Proposal

Refactor `synchronized_received_certificates_from_validator`

## Release Plan

- Nothing to do 
## Links

- [reviewer checklist](https://github.com/linera-io/linera-protocol/blob/main/CONTRIBUTING.md#reviewer-checklist)

### CODE REVIEW

File: linera-core/src/client/mod.rs [hunk 1]
```diff
@@ -1131,98 +1137,140 @@ where
...
+                HandleCertificateResult::New(certificate) => {
+                    certificates.push(certificate);
+                    new_tracker += 1;
+                }
             }
         }
         Ok((remote_node.name, new_tracker, certificates))
     }
 
+    async fn handle_certificate(
```
Comment (afck) [inline linera-core/src/client/mod.rs:1231]: Maybe `check_received_certificate`?

`handle_certificate` elsewhere always means that we update our state based on it, e.g. execute the block, update the chain state, and handle cross-chain messages.

File: linera-core/src/client/mod.rs [hunk 2]
```diff
@@ -1230,11 +1220,9 @@ where
 
     async fn handle_certificate(
         &self,
-        admin_chain: ChainId,
         highest_known_epoch: Epoch,
-        committees: &mut BTreeMap<Epoch, Committee>,
+        committees: &BTreeMap<Epoch, Committee>,
         incoming_certificate: Certificate,
```
Comment (afck) [inline linera-core/src/client/mod.rs:1225]: Why not pass it in by reference, and make the result type a simple enum?

File: linera-core/src/client/mod.rs (hunk 1 above)
Comment (deuszx) [reply linera-core/src/client/mod.rs]: Done.

File: linera-core/src/client/mod.rs (hunk 2 above)
Comment (deuszx) [reply linera-core/src/client/mod.rs]: You're right. Maybe in the original approach (pre-merge to `main`) it made more sense but now it's not true. 

Pass by reference and simplified the enum.

File: linera-core/src/client/mod.rs [hunk 3]
```diff
@@ -1138,78 +1144,106 @@ where
...
-            if !local_response
-                .info
-                .requested_sent_certificate_hashes
-                .is_empty()
-            {
+                })?
+                .info;
+            if !local_response.requested_sent_certificate_hashes.is_empty() {
+                // We've already processed incoming messages for this certificate.
+                // TODO: We verify it's non-empty but how do we make sure we've processed ALL?
```
Comment (deuszx) [inline linera-core/src/client/mod.rs:1165]: @afck can you have a look at this question? I left it in the code so I don't forget.

Comment (afck) [reply linera-core/src/client/mod.rs:1165]: We only requested a single entry, so the collection contains either that or nothing.

File: linera-core/src/client/mod.rs [hunk 4]
```diff
@@ -1138,78 +1144,105 @@ where
...
+            let remote_response = remote_node.handle_chain_info_query(query).await?;
+            let certificate_hash =
+                match remote_response.requested_sent_certificate_hashes.as_slice() {
+                    &[hash] => hash,
+                    [] => {
+                        warn!("Validator didn't have certificate he claimed to have.");
+                        break;
+                    }
+                    _ => {
+                        error!("Validator sent more than one certificate hash for a single block.");
```
Comment (afck) [inline linera-core/src/client/mod.rs:1178]: (We could even return an `InvalidChainInfoResponse` error here: that validator is faulty.)

**Review by afck — APPROVED** [review]

File: linera-core/src/client/mod.rs (hunk 4 above)
Comment (deuszx) [reply linera-core/src/client/mod.rs:1178]: Done in https://github.com/linera-io/linera-protocol/pull/2643/commits/2588a2bb4232f2640c5fac85bcbb4ddf4cac9e02 .

**Pull request merged**